src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java

```diff
@@ -0,0 +1,200 @@
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Dao
+public class ShoppingCartDaoJdbcImpl implements ShoppingCartDao {
+    @Override
+    public Optional<ShoppingCart> getUserShoppingCart(Long userId) {
+        String query = "SELECT * FROM shopping_carts WHERE user_id = ? AND deleted = FALSE";
+        ShoppingCart cart = null;
+
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, userId);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                cart = getCartFromResultSet(resultSet);
+            }
+        } catch (SQLException ex) {
+            throw new DataProcessingException("Can't get shopping cart of user with id = "
+                    + userId, ex);
+        }
+        if (cart != null) {
+            cart.setProducts(getListOfProducts(cart.getId()));
+        }
+        return Optional.ofNullable(cart);
+    }
+
+    @Override
+    public ShoppingCart create(ShoppingCart cart) {
+        String query = "INSERT INTO shopping_carts (user_id) VALUES (?);";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query,
+                    Statement.RETURN_GENERATED_KEYS);
+            statement.setLong(1, cart.getUserId());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                cart.setId(resultSet.getLong(1));
+            }
+        } catch (SQLException ex) {
+            throw new DataProcessingException("Can't create shopping cart for user with id = "
+                    + cart.getUserId(), ex);
+        }
+        insertProductsToShoppingCart(cart);
+        return cart;
+    }
+
+    @Override
+    public Optional<ShoppingCart> get(Long id) {
+        String query = "SELECT * FROM shopping_carts WHERE cart_id = ? AND deleted = FALSE";
+        ShoppingCart cart = null;
+
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                cart = getCartFromResultSet(resultSet);
+            }
+        } catch (SQLException ex) {
+            throw new DataProcessingException("Can't get shopping cart with id = "
+                    + id, ex);
+        }
+
+        if (cart != null) {
+            cart.setProducts(getListOfProducts(cart.getId()));
+        }
+        return Optional.ofNullable(cart);
```

Comment:
why ofNullable = not of?

Reply:
Because in my version cart can be null because of declaration. And if we use Optional.of(null) we will get NPE.